File: Camera/src/main/java/com/example/camera/util/FaceView.java
package com.example.camera.util;

import android.content.Context;
import android.graphics.*;
import android.graphics.drawable.Drawable;
import android.hardware.Camera;
import android.util.AttributeSet;
import android.util.Log;
import android.view.View;
import android.widget.ImageView;

import com.example.camera.R;

/**
 * Created by renlei
 * DATE: 15-11-11
 * Time: 上午11:34
 * Email: renlei0109@yeah.net
 */
public class FaceView extends ImageView {
    private Context mContext;
    private Camera.Face[] mFaces;
    private Matrix mMatrix = new Matrix();
    private boolean mirror;
    private Paint mLinePaint;

    private RectF rectF = new RectF();
    private Drawable mFaceIndicator = null;

    public FaceView(Context context, AttributeSet attrs) {
        super(context, attrs);
        initPaint();
        this.mContext = context;
        mFaceIndicator = mContext.getResources().getDrawable(R.drawable.ic_face_find_2);
    }

    public void setFaces(Camera.Face[] faces) {
        this.mFaces = faces;
        Log.d("Faceview", "invalidate");
//        ((View)getParent()).invalidate();
        invalidate();
        /*postInvalidate();
        invalidate();
        forceLayout();
        requestLayout();*/
    }

    public void clearFaces() {
        mFaces = null;
        invalidate();
    }

    @Override
    protected void onDraw(Canvas canvas) {
//        Log.d("Faceview", "onDraw");
        if (mFaces == null || mFaces.length < 1) {
            return;
        }
        if (mFaces != null) {
            Log.d("renlei", "onDraw" + mFaces.length);
            int id = CameraUtil.getInstance().getCameraId();
            mirror = (id == Camera.CameraInfo.CAMERA_FACING_FRONT);
            canvas.save();
            prepareMatrix();
            mMatrix.postRotate(0); //Matrix.postRotate默认是顺时针
            canvas.rotate(-0);   //Canvas.rotate()默认是逆时针
            for (int i = 0; i < mFaces.length; i++) {
                rectF.set(mFaces[i].rect);
                mMatrix.mapRect(rectF);
                mFaceIndicator.setBounds(Math.round(rectF.left), Math.round(rectF.top),
                        Math.round(rectF.right), Math.round(rectF.bottom));
                mFaceIndicator.draw(canvas);
            }
            canvas.restore();
        }
        super.onDraw(canvas);
    }


    /**
     * <p>Here is the matrix to convert driver coordinates to View coordinates
     * in pixels.</p>
     * <pre>
     * Matrix matrix = new Matrix();
     * CameraInfo info = CameraHolder.instance().getCameraInfo()[cameraId];
     * // Need mirror for front camera.
     * boolean mirror = (info.facing == CameraInfo.CAMERA_FACING_FRONT);
     * matrix.setScale(mirror ? -1 : 1, 1);
     * // This is the value for android.hardware.Camera.setDisplayOrientation.
     * matrix.postRotate(displayOrientation);
     * // Camera driver coordinates range from (-1000, -1000) to (1000, 1000).
     * // UI coordinates range from (0, 0) to (width, height).
     * matrix.postScale(view.getWidth() / 2000f, view.getHeight() / 2000f);
     * matrix.postTranslate(view.getWidth() / 2f, view.getHeight() / 2f);
     * </pre>
     */
    private void prepareMatrix() {
        mMatrix.setScale(mirror ? -1 : 1, 1);
        mMatrix.postRotate(9);
        mMatrix.postScale(getWidth() / 2000f, getHeight() / 2000f);
        mMatrix.postTranslate(getWidth() / 2f, getHeight() / 2f);
    }

    private void initPaint() {
        mLinePaint = new Paint(Paint.ANTI_ALIAS_FLAG);
//		int color = Color.rgb(0, 150, 255);
        int color = Color.rgb(98, 212, 68);
//		mLinePaint.setColor(Color.RED);
        mLinePaint.setColor(color);
        mLinePaint.setStyle(Paint.Style.STROKE);
        mLinePaint.setStrokeWidth(5f);
        mLinePaint.setAlpha(180);
    }
}

File: Camera/src/main/java/com/example/camera/util/FaceView.kt
package com.example.camera.util

import android.content.Context
import android.graphics.*
import android.graphics.drawable.Drawable
import android.hardware.Camera
import android.util.AttributeSet
import android.util.Log
import androidx.appcompat.widget.AppCompatImageView
import com.example.camera.R
import kotlin.math.roundToInt

/**
 * Created by renlei
 * DATE: 15-11-11
 * Time: 上午11:34
 * Email: renlei0109@yeah.net
 */
class FaceView(context: Context, attrs: AttributeSet?) : AppCompatImageView(context, attrs) {

    private var mFaces: Array<Camera.Face>? = null
    private val mMatrix = Matrix()
    private var mirror = false
    private lateinit var mLinePaint: Paint
    private val rectF = RectF()
    private var mFaceIndicator: Drawable? = null

    fun setFaces(faces: Array<Camera.Face>?) {
        mFaces = faces
        Log.d(TAG, "invalidate")
        // ((View)getParent()).invalidate();
        invalidate()
        /*postInvalidate();
        invalidate();
        forceLayout();
        requestLayout();*/
    }

    fun clearFaces() {
        mFaces = null
        invalidate()
    }

    override fun onDraw(canvas: Canvas) {
//        Log.d("Faceview", "onDraw");
        if (mFaces == null || mFaces!!.isEmpty()) {
            return
        }
        if (mFaces != null) {
            Log.d(TAG, "onDraw" + mFaces!!.size)
            val id: Int = CameraInstance.get().cameraId
            mirror = id == Camera.CameraInfo.CAMERA_FACING_FRONT
            canvas.save()
            prepareMatrix()
            mMatrix.postRotate(0f) //Matrix.postRotate默认是顺时针
            canvas.rotate(-0f) //Canvas.rotate()默认是逆时针
            for (i in mFaces!!.indices) {
                rectF.set(mFaces!![i].rect)
                mMatrix.mapRect(rectF)
                mFaceIndicator!!.setBounds(
                    rectF.left.roundToInt(), rectF.top.roundToInt(),
                    rectF.right.roundToInt(), rectF.bottom.roundToInt()
                )
                mFaceIndicator!!.draw(canvas)
            }
            canvas.restore()
        }
        super.onDraw(canvas)
    }

    /**
     *
     * Here is the matrix to convert driver coordinates to View coordinates
     * in pixels.
     * <pre>
     * Matrix matrix = new Matrix();
     * CameraInfo info = CameraHolder.instance().getCameraInfo()[cameraId];
     * // Need mirror for front camera.
     * boolean mirror = (info.facing == CameraInfo.CAMERA_FACING_FRONT);
     * matrix.setScale(mirror ? -1 : 1, 1);
     * // This is the value for android.hardware.Camera.setDisplayOrientation.
     * matrix.postRotate(displayOrientation);
     * // Camera driver coordinates range from (-1000, -1000) to (1000, 1000).
     * // UI coordinates range from (0, 0) to (width, height).
     * matrix.postScale(view.getWidth() / 2000f, view.getHeight() / 2000f);
     * matrix.postTranslate(view.getWidth() / 2f, view.getHeight() / 2f);
    </pre> *
     */
    private fun prepareMatrix() {
        mMatrix.setScale((if (mirror) -1 else 1).toFloat(), 1F)
        mMatrix.postRotate(9F)
        mMatrix.postScale(width / 2000F, height / 2000F)
        mMatrix.postTranslate(width / 2F, height / 2F)
    }

    private fun initPaint() {
        mLinePaint = Paint(Paint.ANTI_ALIAS_FLAG)
        // int color = Color.rgb(0, 150, 255);
        val color = Color.rgb(98, 212, 68)
        // mLinePaint.setColor(Color.RED);
        mLinePaint.color = color
        mLinePaint.style = Paint.Style.STROKE
        mLinePaint.strokeWidth = 5f
        mLinePaint.alpha = 180
    }

    init {
        initPaint()
        mFaceIndicator = context.resources.getDrawable(R.drawable.ic_face_find_2)
    }

    companion object {
        private const val TAG = "FaceView"
    }
}